### ISSUE TITLE
segfault with FlexAttention for causal mask and certain block sizes

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#129911]
### 🐛 Describe the bug

```

# AOT ID: ['0_forward']
from ctypes import c_void_p, c_long
import torch
import math
import random
import os
import tempfile
from math import inf, nan
from torch._inductor.hooks import run_intermediate_hooks
from torch._inductor.utils import maybe_profile
from torch._inductor.codegen.memory_planning import _align as align

from torch import device, empty_strided
from torch._inductor.async_compile import AsyncCompile
from torch._inductor.select_algorithm import extern_kernels
from torch._inductor.codegen.multi_kernel import MultiKernelCall

aten = torch.ops.aten
inductor_ops = torch.ops.inductor
_quantized = torch.ops._quantized
assert_size_stride = torch._C._dynamo.guards.assert_size_stride
empty_strided_cpu = torch._C._dynamo.guards._empty_strided_cpu
empty_strided_cuda = torch._C._dynamo.guards._empty_strided_cuda
reinterpret_tensor = torch._C._dynamo.guards._reinterpret_tensor
alloc_from_pool = torch.ops.inductor._alloc_from_pool
async_compile = AsyncCompile()


# kernel path: /tmp/torchinductor_chilli/aj/cajbrzyvsybzdiw22etwnqyxocghkkgl6lo4eb77aynfr3p7pkup.py
# Source Nodes: [flex_attention], Original ATen: []
# flex_attention => flex_attention
import triton
import triton.language as tl
from triton.compiler.compiler import AttrsDescriptor

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, instance_descriptor, DeviceProperties

@triton_heuristics.template(
    num_stages=2,
    num_warps=8,
    triton_meta={'signature': {0: '*fp16', 1: '*fp16', 2: '*fp16', 3: '*fp32', 4: '*i32', 5: '*i32', 6: '*fp16'}, 'device': DeviceProperties(type='cuda', index=0, cc=90, major=9, regs_per_multiprocessor=65536, max_threads_per_multi_processor=2048, multi_processor_count=132), 'constants': {}, 'configs': [AttrsDescriptor(divisible_by_16=(0, 1, 2, 3, 4, 5, 6), equal_to_1=())]},
    inductor_meta={'kernel_name': 'triton_tem_fused_0', 'backend_hash': 'E85CA86B443BE391E364C5E50BBA3418FF1E341B8977AF416D4AEE4BB2C023D0', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': False, 'force_disable_caches': False, 'dynamic_scale_rblock': True, 'max_autotune': True, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False},
)
@triton.jit
def triton_tem_fused_0(arg_Q, arg_K, arg_V, arg_LSE, arg_SPARSE_KV_NUM_BLKS, arg_SPARSE_KV_IDX, out_ptr0):
    BLOCK_M : tl.constexpr = 128
    BLOCK_N : tl.constexpr = 128
    BLOCK_DMODEL : tl.constexpr = 64
    SCORE_MOD_IS_LINEAR : tl.constexpr = False
    ROWS_GUARANTEED_SAFE : tl.constexpr = False
    OUTPUT_LOGSUMEXP : tl.constexpr = True
    SPARSE_Q_BLOCK_SIZE : tl.constexpr = 128
    SPARSE_KV_BLOCK_SIZE : tl.constexpr = 128
    Q = arg_Q
    K = arg_K
    V = arg_V
    LSE = arg_LSE
    SPARSE_KV_NUM_BLKS = arg_SPARSE_KV_NUM_BLKS
    SPARSE_KV_IDX = arg_SPARSE_KV_IDX

    # Sub notation for this kernel:
    #
    # Q: Query, K: Key, V: Value
    # M: Number of queries, N: Number of keys/values, D: Model dimension
    # z: Batch size, h: Number of heads, m: Number of queries per head, k: Number of keys per head
    # (Modifiable) Config options:
    # BLOCK_M: The thread block size across the seqlen dim of Q.
    # BLOCK_N: Iterate over BLOCK_N across the seqlen dim of K/V in each thread block.
    # SCORE_MOD_IS_LINEAR: Is the score modifier linear? If so, we can lift the
    # change of base out of the loop
    # ROWS_GUARANTEED_SAFE: Is it guaranteed that at least one value in each row
    # is not masked out? If so, we can skip an extra safety check
    # OUTPUT_LOGSUMEXP: We only need to store the logsumexp if we require grad
    #
    # The following SPARSE_* is defined in the block sparse mask grid, rather than the thread block grid.
    # SPARSE_KV_NUM_BLKS: The number of unmasked K/V blocks for each query.
    # SPARSE_KV_IDX: The indices of unmasked K/V blocks for each query.

    tl.static_assert(SPARSE_Q_BLOCK_SIZE >= BLOCK_M and SPARSE_Q_BLOCK_SIZE % BLOCK_M == 0)
    tl.static_assert(SPARSE_KV_BLOCK_SIZE >= BLOCK_N and SPARSE_KV_BLOCK_SIZE % BLOCK_N == 0)

    # Define Q Strides
    stride_qz = 4194304
    stride_qh = 524288
    stride_qm = 64
    stride_qk = 1
    # Define K Strides
    stride_kz = 4194304
    stride_kh = 524288
    stride_kn = 64
    stride_kk = 1
    # Define V Strides
    stride_vz = 4194304
    stride_vh = 524288
    stride_vk = 64
    stride_vn = 1

    Z = 16
    H = 8
    Q_LEN = 8192
    KV_LEN = 8192

    qk_scale = 1.0
    MATMUL_PRECISION = Q.dtype.element_ty

    start_m = tl.program_id(0)
    off_z = tl.program_id(1) // H
    off_h = tl.program_id(1) % H

    q_offset = off_z * stride_qz + off_h * stride_qh
    k_offset = off_z * stride_kz + off_h * stride_kh
    v_offset = off_z * stride_vz + off_h * stride_vh

    SPARSE_Z = 1
    SPARSE_H = 1

    sparse_idx_z = off_z % SPARSE_Z
    sparse_idx_h = off_h % SPARSE_H

    SPARSE_Q_MULTIPLE: tl.constexpr = (SPARSE_Q_BLOCK_SIZE // BLOCK_M)
    SPARSE_KV_MULTIPLE: tl.constexpr = (SPARSE_KV_BLOCK_SIZE // BLOCK_N)

    SPARSE_Q_BLOCK_CNT: tl.constexpr = Q_LEN // SPARSE_Q_BLOCK_SIZE
    SPARSE_KV_BLOCK_CNT: tl.constexpr = KV_LEN // SPARSE_KV_BLOCK_SIZE

    # SPARSE_KV_IDX and SPARSE_KV_NUM_BLKS are always contiguous.
    sparse_hz_offset = sparse_idx_z * SPARSE_H + sparse_idx_h
    sparse_kv_num_blks_offset = sparse_hz_offset * SPARSE_Q_BLOCK_CNT + start_m // SPARSE_Q_MULTIPLE
    sparse_kv_idx_offset = sparse_hz_offset * SPARSE_Q_BLOCK_CNT * SPARSE_KV_BLOCK_CNT + (start_m // SPARSE_Q_MULTIPLE) * SPARSE_KV_BLOCK_CNT  # noqa: B950
    kv_indices = SPARSE_KV_IDX + sparse_kv_idx_offset
    kv_start = tl.load(kv_indices) * SPARSE_KV_BLOCK_SIZE # first kv block we're loading
    sparse_kv_num_blocks = tl.load(SPARSE_KV_NUM_BLKS + sparse_kv_num_blks_offset)

    Q_block_ptr = tl.make_block_ptr(
        base=Q + q_offset,
        shape=(Q_LEN, BLOCK_DMODEL),
        strides=(stride_qm, stride_qk),
        offsets=(start_m * BLOCK_M, 0),
        block_shape=(BLOCK_M, BLOCK_DMODEL),
        order=(1, 0)
    )
    K_block_ptr = tl.make_block_ptr(
        base=K + k_offset,
        shape=(BLOCK_DMODEL, KV_LEN),
        strides=(stride_kk, stride_kn),
        offsets=(0, kv_start),
        block_shape=(BLOCK_DMODEL, BLOCK_N),
        order=(0, 1)
    )
    V_block_ptr = tl.make_block_ptr(
        base=V + v_offset,
        shape=(KV_LEN, BLOCK_DMODEL),
        strides=(stride_vk, stride_vn),
        offsets=(kv_start, 0),
        block_shape=(BLOCK_N, BLOCK_DMODEL),
        order=(1, 0)
    )
    # initialize offsets
    offs_m = start_m * BLOCK_M + tl.arange(0, BLOCK_M)
    offs_n = kv_start + tl.arange(0, BLOCK_N)
    # initialize pointer to m and l
    m_i = tl.zeros([BLOCK_M], dtype=tl.float32) - float("inf")
    l_i = tl.zeros([BLOCK_M], dtype=tl.float32)
    acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)

    q = tl.load(Q_block_ptr)
    if SCORE_MOD_IS_LINEAR:
        qk_scale *= 1.44269504
    q = (q * qk_scale).to(MATMUL_PRECISION)

    # loop over k, v and update accumulator
    lo = 0
    hi = sparse_kv_num_blocks * SPARSE_KV_MULTIPLE

    for start_n in range(0, hi):
        # -- load k --
        k = tl.load(K_block_ptr)
        # -- compute qk ---
        qk = tl.dot(q, k)
        # ~~~~~~~~~~~~~~~~~~~ Apply score modification  ~~~~~~~~~~~~~~~~~~~
        m = offs_m[:, None]
        n = offs_n[None, :]
        tmp0 = (m) >= (n)
        tmp1 = float("-inf")
        tmp2 = tl.where(tmp0, (qk), tmp1)
        post_mod_scores = tmp2

        # TODO: In the case that score_mod is linear, this can be LICMed
        if not SCORE_MOD_IS_LINEAR:
            post_mod_scores *= 1.44269504
        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

        # -- compute scaling constant ---
        m_ij = tl.maximum(m_i, tl.max(post_mod_scores, 1))

        alpha = tl.math.exp2(m_i - m_ij)
        p = tl.math.exp2(post_mod_scores - m_ij[:, None])
        if not ROWS_GUARANTEED_SAFE:
            masked_out_rows = (m_ij == float("-inf"))
            alpha = tl.where(masked_out_rows, 0, alpha)
            p = tl.where(masked_out_rows[:, None], 0, p)

        # NB: l_i update is pulled up here since it's a bit faster
        # NB: For headdim=256, it's faster to move it back down to after m_i =
        # m_ij
        l_i = l_i * alpha + tl.sum(p, 1)
        # # -- scale and update acc --
        acc = acc * alpha[:, None]
        v = tl.load(V_block_ptr)
        acc = tl.dot(p.to(MATMUL_PRECISION), v, acc)

        # -- update m_i
        m_i = m_ij

        # update pointers
        indices_idx = start_n // SPARSE_KV_MULTIPLE

        cur_block = tl.load(kv_indices + indices_idx)
        next_block = tl.load(kv_indices + indices_idx + 1)
        needs_jump = (start_n + 1) % SPARSE_KV_MULTIPLE == 0
        jump_to_block = (next_block - cur_block ) * SPARSE_KV_BLOCK_SIZE - (SPARSE_KV_MULTIPLE - 1) * BLOCK_N

        offset = jump_to_block * needs_jump + (1 - needs_jump) * BLOCK_N

        V_block_ptr = tl.advance(V_block_ptr, (offset, 0))
        K_block_ptr = tl.advance(K_block_ptr, (0, offset))

        offs_n = offs_n + offset

    # Store output and logsumexp
    acc = acc / l_i[:, None]
    idx_z = tl.program_id(1) // H
    idx_h = tl.program_id(1) % H
    idx_m = offs_m[:, None]
    idx_d = tl.arange(0, BLOCK_DMODEL)[None, :]

    mask = idx_m < Q_LEN
    # TODO generalize and add proper mask support
    xindex = idx_d + (64*idx_m) + (524288*idx_h) + (4194304*idx_z)
    tl.store(out_ptr0 + (tl.broadcast_to(xindex, acc.shape)), acc, mask)

    # TODO dont want to write this if we dont require grad
    if OUTPUT_LOGSUMEXP:
        off_hz = tl.program_id(1)
        l_ptrs = LSE + off_hz * Q_LEN + offs_m
        lse = m_i + tl.math.log2(l_i)
        tl.store(l_ptrs, lse)

import torch._inductor.kernel.flex_attention
meta0 = {'BLOCK_M': 128, 'BLOCK_N': 128, 'BLOCK_DMODEL': 64, 'SCORE_MOD_IS_LINEAR': False, 'ROWS_GUARANTEED_SAFE': False, 'OUTPUT_LOGSUMEXP': True, 'SPARSE_Q_BLOCK_SIZE': 128, 'SPARSE_KV_BLOCK_SIZE': 128}
from torch._C import _cuda_getCurrentRawStream as get_raw_stream

import triton
import triton.language as tl
from torch._inductor.runtime.triton_heuristics import grid, split_scan_grid, start_graph, end_graph


async_compile.wait(globals())
del async_compile

def call(args):
    primals_1, primals_2, primals_3, primals_4, primals_5, primals_6, primals_7 = args
    args.clear()
    assert_size_stride(primals_1, (16, 8, 8192, 64), (4194304, 524288, 64, 1))
    assert_size_stride(primals_2, (16, 8, 8192, 64), (4194304, 524288, 64, 1))
    assert_size_stride(primals_3, (16, 8, 8192, 64), (4194304, 524288, 64, 1))
    assert_size_stride(primals_4, (1, 1, 64), (64, 64, 1))
    assert_size_stride(primals_5, (1, 1, 64, 64), (4096, 4096, 64, 1))
    assert_size_stride(primals_6, (1, 1, 64), (64, 64, 1))
    assert_size_stride(primals_7, (1, 1, 64, 64), (4096, 4096, 64, 1))
    buf0 = empty_strided_cuda((16, 8, 8192), (65536, 8192, 1), torch.float32)
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0)
        buf1 = empty_strided_cuda((16, 8, 8192, 64), (4194304, 524288, 64, 1), torch.float16)
        # Source Nodes: [flex_attention], Original ATen: []
        stream0 = get_raw_stream(0)
        triton_tem_fused_0.run(primals_1, primals_2, primals_3, buf0, primals_4, primals_5, buf1, grid=torch._inductor.kernel.flex_attention.flex_attention_grid(16, 8, 8192, 64, meta0), stream=stream0)
    return (buf1, primals_1, primals_2, primals_3, primals_4, primals_5, primals_6, primals_7, buf1, buf0, )


def benchmark_compiled_module(times=10, repeat=10):
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    primals_1 = rand_strided((16, 8, 8192, 64), (4194304, 524288, 64, 1), device='cuda:0', dtype=torch.float16)
    primals_2 = rand_strided((16, 8, 8192, 64), (4194304, 524288, 64, 1), device='cuda:0', dtype=torch.float16)
    primals_3 = rand_strided((16, 8, 8192, 64), (4194304, 524288, 64, 1), device='cuda:0', dtype=torch.float16)
    primals_4 = rand_strided((1, 1, 64), (64, 64, 1), device='cuda:0', dtype=torch.int32)
    primals_5 = rand_strided((1, 1, 64, 64), (4096, 4096, 64, 1), device='cuda:0', dtype=torch.int32)
    primals_6 = rand_strided((1, 1, 64), (64, 64, 1), device='cuda:0', dtype=torch.int32)
    primals_7 = rand_strided((1, 1, 64, 64), (4096, 4096, 64, 1), device='cuda:0', dtype=torch.int32)
    fn = lambda: call([primals_1, primals_2, primals_3, primals_4, primals_5, primals_6, primals_7])
    return print_performance(fn, times=times, repeat=repeat)


if __name__ == "__main__":
    from torch._inductor.wrapper_benchmark import compiled_module_main
    compiled_module_main('None', benchmark_compiled_module)
```

This turns into a segfault on H100.

```
scheduleRemainingToLastStage (forOp=..., schedule=..., afterPrologue=..., afterPrologue@entry=..., numStages=numStages@entry=2) at ../../../lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:893
```

### Versions

nightly.

cc @bertmaher @int3 @davidberard98 @nmacchioni @chenyang78 @embg

**Comment by Chillee** [comment]
i can confirm this is identical to https://github.com/pytorch/pytorch/issues/129625
